src/cpp/r/RSexp.cpp

```diff
@@ -838,6 +896,164 @@ bool inherits(SEXP object, const char* S3Class)
    return Rf_inherits(object, S3Class);
 }
 
+std::set<std::string> makeNsePrimitives()
+{
+   std::set<std::string> nsePrimitives;
+   nsePrimitives.insert("quote");
+   nsePrimitives.insert("substitute");
+   nsePrimitives.insert("match.call");
+   nsePrimitives.insert("library");
+   nsePrimitives.insert("require");
+   nsePrimitives.insert("enquote");
+   nsePrimitives.insert("bquote");
+   nsePrimitives.insert("expression");
+   nsePrimitives.insert("evalq");
+   nsePrimitives.insert("subset");
+   return nsePrimitives;
+}
+
+const std::set<std::string>& nsePrimitives()
+{
+   static const std::set<std::string> set = makeNsePrimitives();
+   return set;
+}
+
+// Attempts to find calls to functions which perform NSE.
+bool maybePerformsNSEImpl(SEXP node,
+                          const std::set<std::string>& nsePrimitives)
+{
+   if (TYPEOF(node) == LANGSXP)
+   {
+      // Check to see if this is a call to a function typically
+      // used with non-standard evaluation
+      SEXP head = CAR(node);
+      if (TYPEOF(head) == SYMSXP && nsePrimitives.count(CHAR(PRINTNAME(head))))
+         return true;
+      
+      // Iterate over all other elements in the pairlist and check to see if
+      // they make calls to NSE primitives.
+      SEXP tail = CDR(node);
+      head = CAR(tail);
+      while (tail != R_NilValue)
+      {
+         if (TYPEOF(head) == LANGSXP && maybePerformsNSEImpl(head, nsePrimitives))
+            return true;
+         
+         tail = CDR(tail);
+         head = CAR(tail);
+      }
+   }
+   
+   return false;
+}
+
+bool maybePerformsNSE(SEXP function)
+{
+   if (!Rf_isFunction(function))
+      return false;
+   
+   if (Rf_isPrimitive(function))
+      return false;
+   
+   return maybePerformsNSEImpl(
+            BODY_EXPR(function),
+            nsePrimitives());
+}
+
+// NOTE: Uses `R_lsInternal` which throws error if a non-environment is
+// passed; we therefore perform this validation ourselves before calling
+// `R_lsInternal`.
```

Comment:
Again we can use executeSafely to eliminate the problem of thrown errors in R API functions.